src/shared/conf-parser.c

```diff
@@ -568,21 +590,127 @@ int config_parse_many(
                 void *userdata,
                 Hashmap **ret_stats_by_path) {
 
-        _cleanup_strv_free_ char **dropin_dirs = NULL;
         _cleanup_strv_free_ char **files = NULL;
-        const char *suffix;
         int r;
 
-        suffix = strjoina("/", dropin_dirname);
-        r = strv_extend_strv_concat(&dropin_dirs, (char**) conf_file_dirs, suffix);
+        assert(conf_file_dirs);
+        assert(dropin_dirname);
+        assert(sections);
+        assert(table);
+
+        r = config_get_dropin_files(conf_file_dirs, dropin_dirname, &files);
+        if (r < 0)
+                return r;
+
+        return config_parse_many_files(conf_files, files, sections, lookup, table, flags, userdata, ret_stats_by_path);
+}
+
+static int config_get_stats_by_path_one(
+                const char* conf_file,
+                const char* const* conf_file_dirs,
+                Hashmap *stats_by_path) {
+
+        _cleanup_strv_free_ char **files = NULL;
+        _cleanup_free_ char *dropin_dirname = NULL;
+        struct stat st;
+        int r;
+
+        assert(conf_file);
+        assert(conf_file_dirs);
+        assert(stats_by_path);
+
+        /* Unlike config_parse(), this does not support stream. */
+
+        r = path_extract_filename(conf_file, &dropin_dirname);
         if (r < 0)
                 return r;
+        if (r == O_DIRECTORY)
+                return -EINVAL;
+
+        if (!strextend(&dropin_dirname, ".d"))
+                return -ENOMEM;
 
-        r = conf_files_list_strv(&files, ".conf", NULL, 0, (const char* const*) dropin_dirs);
+        r = config_get_dropin_files(conf_file_dirs, dropin_dirname, &files);
         if (r < 0)
                 return r;
 
-        return config_parse_many_files(conf_files, files, sections, lookup, table, flags, userdata, ret_stats_by_path);
+        /* First read the main config file. */
+        r = RET_NERRNO(stat(conf_file, &st));
+        if (r >= 0) {
+                r = hashmap_put_stats_by_path(&stats_by_path, conf_file, &st);
+                if (r < 0)
+                        return r;
+        } else if (r != -ENOENT)
+                return r;
+
+
+        /* Then read all the drop-ins. */
+        STRV_FOREACH(fn, files) {
+                if (stat(*fn, &st) < 0) {
+                        if (errno == ENOENT)
+                                continue;
```

Comment:
There's probably a good reason but I don't entirely understand why we need to ignore ENOENT here. Can you eplxain that (and maybe add a comment)?

Reply:
TOCTTOU. i.e. you enumerate the file with readddir()/getdents(), but when you try to stat it, it's already gone. All across our codebase we should thus handle cases gracefully where after seeing a dir in readdir() the stat() on it fails with ENOENT

Reply:
Right, the file may be removed after readdir().